File: src/main/java/algo/sequence/LongestSubstringWithAtLeastKRepeatingCharacters_395.java
package algo.sequence;

import java.util.HashMap;

public class LongestSubstringWithAtLeastKRepeatingCharacters_395 {
  public static int longestSubstring(String s, int k) {
    // calculate frequency
    var freq = new HashMap<Character, Integer>();
    for (int i = 0; i < s.length(); i++) {
      freq.put(s.charAt(i), freq.getOrDefault(s.charAt(i), 0) + 1);
    }

    // find the character whose presents are less than k
    var delimeter = "";
    for (var entry : freq.entrySet()) {
      if (entry.getValue() < k) {
        delimeter = entry.getKey() + "";
      }
    }

    // we find a qualified string and return the length
    if ("".equals(delimeter)) {
      return s.length();
    }

    var max = 0;
    for (var segment : s.split(delimeter)) {
      max = Math.max(max, longestSubstring(segment, k));
    }

    return max;
  }

  public static void main(String[] args) {
    System.out.println(longestSubstring("aaabb", 3));
    System.out.println(longestSubstring("ababbc", 2));
  }
}

File: src/main/kotlin/algo/sequence/LongestSubstringWithAtLeastKRepeatingCharacters_395.kt
package algo.sequence

class KtLongestSubstringWithAtLeastKRepeatingCharacters_395 {
    fun longestSubstring(s: String, k: Int): Int {
        // chars whose occurrence less than k can be excluded
        val freq: MutableMap<Char, Int> = mutableMapOf()
        s.forEach { ch ->
            freq[ch] = freq.computeIfAbsent(ch) { 0 } + 1
        }

        val delimiters = freq.filter { it.value < k }.map { it.key }.toTypedArray()
        if (delimiters.isEmpty()) {
            return s.length
        }

        var maxLen = 0
        s.split(Regex(delimiters.joinToString("|")))
            .filter { it.isNotEmpty() }
            .forEach { sub ->
            maxLen = maxLen.coerceAtLeast(longestSubstring(sub, k))
        }

        return maxLen
    }
}

fun main() {
    val substring = KtLongestSubstringWithAtLeastKRepeatingCharacters_395()
    println(substring.longestSubstring("aaabb", 3))
    println(substring.longestSubstring("ababbc", 2))
}